### ISSUE TITLE
Guides - Update Developer Guide to Reflect New Process

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#2863]
It's a new year and we've introduced a new "develop" branch, which is now the default. We need to update the "Branching Strategy" and "Making Releases" docs to explain the purpose of the "develop" branch:
- http://guides.dataverse.org/en/4.2.3/developers/branching-strategy.html
- http://guides.dataverse.org/en/4.2.3/developers/making-releases.html

Top of mind thoughts:
- The "develop" branch represents all the code that will go in the next release (4.2.4 as of this writing).
- Code changes should be made in branches and tested before they are merged into "develop".
- We can probably replace references to "release branches" and explain that "develop" is what eventually gets merged into master.

Questions:
- Should pull requests be made against "develop"? I think so... they have to be made against **somewhere**... Where should pull requests be tested? We don't have fancy integration-tests-per-pull-request like @openscholar does in https://travis-ci.org/openscholar/openscholar/builds . See also #2746 and https://github.com/IQSS/dataverse-puppet/issues/10
- Do pull requests require a Request For Integration (RFI)?
- Should documentation changes (such as updating the Dev Guide, which is what this issue is about) be made in a branch and then merged after approved or is it ok for them to be made directly in "develop" if you have access to do so?
- Is it safe to say that we are basically adopting the "git-flow" model? See http://nvie.com/posts/a-successful-git-branching-model/
- Where should the top part vs. bottom part of https://github.com/IQSS/dataverse/blob/develop/PULL_REQUEST_TEMPLATE.md be explained? In the template itself or in the dev guide?

In case it's helpful, at the time I wrote the [old/current branching strategy doc](http://guides.dataverse.org/en/4.2.3/developers/branching-strategy.html), we were talking a lot about a pipeline where there's handoff from design to dev to QA and a three week development cycle. Here's [the picture](https://plus.google.com/+PhilipDurbin/posts/izE8GUEKASD) I drew back then:

![devcycle](https://cloud.githubusercontent.com/assets/21006/12314127/27980242-ba3c-11e5-8e9c-6ac3a03e654c.jpg)

**Comment by bencomp** [comment]
@pdurbin Is there any consensus on the branching strategy at IQSS, or is this just your idea and up for discussion? If PRs keep being closed for making them against branches that are closed because the branching strategy may change, I may very well lose interest in contributing. Moreover, if it's just you (and for just a tiny bit, me) who determine(s) how developers should write code and commit it, what are the odds that this new strategy is implemented and adhered to?

I would like to have this discussion with all developers, because this single issue on GitHub is easily missed by many. Could you (or rather the team lead(s)) start this wider discussion on the [community list](https://groups.google.com/forum/#!forum/dataverse-community)?

**Comment by pdurbin** [comment]
@bencomp it was nice chatting with you at http://irclog.iq.harvard.edu/dataverse/2016-01-14 and I hope I answered most of your questions! We love our contributors! Please stay tuned!

**Comment by bencomp** [comment]
You did answer my main question about consensus on IRC (I admit I was a bit frustrated when I wrote the above comment).
It would be helpful to explain the reasons for the (changed) branching strategy and refer to the team decision. Perhaps add some background about how changes were pushed to master and what problems you ran into?

Regarding your questions:

- yes, pull requests should be made against a persistent branch (that should prevent having to reopen pull requests when a branch is closed…)
- a pull request should explain what was changed and for what reason(s). Referring to issues seems good practice to me. If "fixes #issue" is in the subject, merging the PR also closes the issue. Should a Request for Integration include more?
- I don't know what is best for (user) documentation, it may depend on the context of documentation changes. If a change accompanies a feature, just put it in the same feature branch and/or pull request. If making small changes to formulations or correcting typos, I don't mind pushing them directly to the develop branch.

**Comment by pdurbin** [comment]
Please wait for something more official but my take on the reasons for the change are:

- We want code to be better tested in a branch before it gets merged into the branch that represents the next release.
- We want developers to know that they can always branch from "develop" when starting work on a new feature or bug fix.

That said, we're still discussing. And really, we need to experiment a bit with any new strategy and get our hands dirty before we can say if we're happy with the change. I'm well aware that changing strategies is hard on contributors, so please pardon our dust!

I started working on a pull request section yesterday and I'm reminded of this quote I wrote in the [contributing file](https://github.com/IQSS/dataverse/blob/develop/CONTRIBUTING.md), which is still what I want:

"Before you start coding, please reach out to us either on our dataverse-community Google Group, IRC, or via support@dataverse.org to make sure the effort is well coordinated and we avoid merge conflicts."

I was starting to think that the emphasis of whatever doc should be not what the branching strategy is but rather clear advice for contributors on how to make your pull request such that it's more likely to be merged.

**Comment by pdurbin** [comment]
@kcondon as we've discussed I made a very small change in b2fcf4d to top of the following pages to note that the branching strategy is under review and that this issue is the one to follow for updates:

- http://guides.dataverse.org/en/4.2.3/developers/branching-strategy.html
- http://guides.dataverse.org/en/4.2.3/developers/making-releases.html

I just did a build and these changes are live so I'm passing it to QA. As agreed, if it looks fine let's move this issue to the 4.3 milestone so we can further improve the docs.

**Comment by kcondon** [comment]
OK, verified the above changes. Marking the follow on work described by this ticket for v4.3.

**Comment by mheppler** [comment]
I've expanded this ticket to be updating the Developer Guide as a whole, since other pages also needed updated.

- [ ] Branching Strategy
- [ ] Testing
- [ ] Making Releases
- [ ] Documentation

**Comment by mheppler** [comment]
Assigning @scolapasta this ticket, as the Lead Developer, to take the lead on bringing this home.

**Comment by pdurbin** [comment]
Speaking of the [CONTRIBUTING.md](https://github.com/IQSS/dataverse/blob/develop/CONTRIBUTING.md) file, #2594 is related and I made a variety of improvements in b622805 to CONTRIBUTING.md in a new branch called 2594-contributing. I think the pull requests section of CONTRIBUTING.md doesn't necessarily need to change. What's important is to have persistent branch (as @bencomp mentioned) for pull requests to be made against. We have that now already with the "develop" branch, and I made it the default branch so pull requests will naturally be made against "develop". Anyway, it might make sense to work on this issue and #2594 at the same time but it probably isn't strictly necessary.

**Comment by pdurbin** [comment]
As I [mentioned on Twitter](https://twitter.com/philipdurbin/status/710986867437670401) @joeyh ran through setting up a development enviroment at [SpinachCon](https://groups.google.com/forum/#!topic/dataverse-community/G6cyCAv1hbY) on Friday (thanks!). Here are the notes I took.

- Is 6 GB of storage enough? 512 MB RAM. Debian Stable (Jesse, 8.2) on a VM in the cloud. Used Ubuntu instructions.
- Text quite small on [feature request process diagram](https://github.com/IQSS/dataverse/blob/v4.3/doc/sphinx-guides/source/img/feature-request-process.png). Hard to read. Diagram is complicated, maybe put "Bug Reports/Issues" section above it. He's a developer. Just wants to fix a bug.
- wget of glassfish is not secure. Cert problem at https://download.java.net/glassfish/4.1/release/glassfish-4.1.zip if you replace HTTP with HTTPS, which he recommends throughout.
- Ubuntu docs just kind of end. 
- What is Solr? Explain what it is.
- Consider taking "harvard" out of default smtp server.
- Postgres password. Explain you have to do something before running the installer.
- "Have you built the war file?" warning. Dev guide doesn't say to.
- Couldn't build the war file due to download problems (timeouts?) with Maven. Grabbed the latest release instead.
- Couldn't deploy war file because Java 7 was the default. Emphasize need to make Java 8 the default.

To sum up, we weren't able to quickly get the installer to work on his VM but we came close and captured some good feedback along the way.

**Comment by djbrooke** [comment]
Removing from 4.5. Will be re-prioritized for a future release. 

This will be updated at the same time as #2594

**Comment by djbrooke** [comment]
Additional feedback from the community, some likely covered elsewhere: 

1. Provide the minimum and recommended configuration of the server.
2. Handles
3. Harvesting
4. Two Ravens graphic tutorial

**Comment by pdurbin** [comment]
This morning @djbrooke asked if it's possible to merge a release into the master branch via the GUI but from a quick look, I don't think it's possible. @kcondon is more of an authority on this than I am since when 4.0 was in development he regularly merged from the master branch to the beta branch (for example Beta 13 in #1482). I'm leaving this comment here because http://guides.dataverse.org/en/4.5/developers/making-releases.html#merge-release-branch no longer reflects what we actually do now that we have a "develop" branch.

Anyway, below are screenshots from my attempt to merge v4.5 into master via the GUI. @djbrooke please advise if you'd like me to train someone how to do this via the command line instead.

![screen shot 2016-09-01 at 12 07 11 pm](https://cloud.githubusercontent.com/assets/21006/18178286/bcf30e16-704b-11e6-93df-022fc0756ef4.png)

![screen shot 2016-09-01 at 12 07 27 pm](https://cloud.githubusercontent.com/assets/21006/18178285/bcf1132c-704b-11e6-9e6d-bd2cf7410f09.png)

![screen shot 2016-09-01 at 12 07 46 pm](https://cloud.githubusercontent.com/assets/21006/18178284/bceedcc4-704b-11e6-95cb-2add66f461df.png)

![screen shot 2016-09-01 at 12 08 08 pm](https://cloud.githubusercontent.com/assets/21006/18178287/bcf5a446-704b-11e6-9b7d-835bb5bbd661.png)

**Comment by pdurbin** [comment]
@mheppler and I were just looking at #3113 and a consequence of our new-as-of-2016 process of "let's use pull requests" (which I'm a big fan of, by the way) is that we generate many more branches. I'm in the habit of deleting branches I've created after they've been merged. If we'd like all developers to do this, we should document this in the dev guide.

**Comment by pdurbin** [comment]
I mentioned to @michbarsinai that developers are now asked to use the special "connects to" syntax to associate pull requests will issues. This should be added to the Dev Guide and the PULL_REQUEST_TEMPLATE.md. Here are the docs from Waffle explaining the syntax: https://github.com/waffleio/waffle.io/wiki/FAQs#prs-and-issues

**Comment by pdurbin** [comment]
I just created pull request #3734 and moved this issue to Code Review at https://waffle.io/IQSS/dataverse

**Comment by pdurbin** [comment]
@mheppler provided some better wording in PULL_REQUEST_TEMPLATE.md which I just added in d86777b

**Comment by dlmurphy** [comment]
I've made some tiny typo fixes to the updated documents and everything is now looking great to me!

**Comment by pdurbin** [comment]
@dlmurphy thanks for catching those typos!

**Comment by pdurbin** [comment]
This shouldn't be a blocker for merging pull request #3734 but I did just solicit feedback from external folks at https://groups.google.com/d/msg/dataverse-dev/6qQVpb8c3zg/r783YQxkBAAJ

I'm moving this to QA at https://waffle.io/IQSS/dataverse
